src/raftstore/store/worker/apply.rs

```diff
@@ -2040,180 +2257,332 @@ impl Runner {
+                                        notify_stale_command(&delegate.tag, delegate.term, cmd);
+                                    }
+                                    delegate.pending_cmds.set_conf_change(cmd);
+                                } else {
+                                    delegate.pending_cmds.append_normal(cmd);
+                                }
+                            }
+
+                            box future::ok(delegate)
+                        }
                     }
-                    delegate.pending_cmds.set_conf_change(cmd);
-                } else {
-                    delegate.pending_cmds.append_normal(cmd);
-                }
-            }
-        }
-        APPLY_PROPOSAL.observe(propose_num as f64);
+                })
+            })
+            .forget();
     }
 
     fn handle_registration(&mut self, s: Registration) {
         let peer_id = s.id;
         let region_id = s.region.get_id();
         let term = s.term;
-        let delegate =
-            ApplyDelegate::from_registration(Arc::clone(&self.db), Arc::clone(&self.raft_db), s);
+        let delegate = ApplyDelegate::from_registration(
+            Arc::clone(&self.db),
+            Arc::clone(&self.raft_db),
+            s,
+            self.apply_scheduler.clone(),
+            self.notifier.clone(),
+            Arc::clone(&self.host),
+            Arc::clone(&self.importer),
+            self.use_delete_range,
+        );
         info!(
-            "{} register to apply delegates at term {}",
-            delegate.tag, delegate.term
+            "{} peer {} register to apply pool at term {}",
+            delegate.tag, peer_id, delegate.term
         );
-        if let Some(mut old_delegate) = self.delegates.insert(region_id, Some(delegate)) {
-            let old_delegate = old_delegate.as_mut().unwrap();
-            assert_eq!(old_delegate.id, peer_id);
-            old_delegate.term = term;
-            old_delegate.clear_all_commands_as_stale();
+
+        let (tx, rx) = unbounded();
+        if let Some(tx) = self.region_task_senders.insert(region_id, tx) {
+            tx.unbounded_send(RegionTask::suicide(region_id, term))
+                .unwrap_or_else(|_| {
+                    error!("Dangling message sender for region {}", region_id);
+                });
         }
+
+        Self::spawn_region(&self.apply_pool, delegate, rx);
     }
 
     fn handle_destroy(&mut self, d: Destroy) {
         // Only respond when the meta exists. Otherwise if destroy is triggered
         // multiple times, the store may destroy wrong target peer.
-        if let Some(meta) = self.delegates.remove(&d.region_id) {
-            let mut meta = meta.unwrap();
-            info!("{} remove from apply delegates", meta.tag);
-            meta.destroy();
-            self.notifier.send(TaskRes::Destroy(meta)).unwrap();
+        if let Some(tx) = self.region_task_senders.remove(&d.region_id) {
```

Comment:
Should we leave something to tell the target is borrowed or stale when new proposals come?

Reply:
There will no proposal for the source region after it has been borrowed.

Reply:
It is guaranteed by merge.